New biomol fields (#400)
Materials-Consortia/OPTIMADE: optimade.rst

Appendices
==========

Domain Specific Fields
----------------------

The fields below are all optional and are only used within specific research fields.

Every field has a standard domain-specific prefix.

> biomol_chains
> ~~~~~~

Reviewer: ```suggestion
biomol_chains
~~~~~~~~~~~
```
I think the line with tildes is supposed to be as long as the title

     If supported, only a subset of the filter features MAY be supported.
   - **name**: The residue name
   - **number**: The residue number according to source notation.
   - **insertion_code**: The residue insertion code. It MUST NOT be longer than 1 character. It MAY be null.
   - **sites**: A list of integers referring to the index of :field:`cartesian_site_positions`, that belong to this residue.
     The list SHOULD NOT be empty. The index of the first site is 0.
   - There MUST NOT be two or more residues with the same integer in :property:`sites`.
   - All :property:`name` and :property:`insertion_code` values SHOULD be in capital letters.

- **Examples**:

.. code:: jsonc
  {
    "biomol_residues":[
      {
        "name": "PHE",
	      "number": 17,
	      "insertion_code": null,
        "sites":[0,1,2,3, ...]
      },
      {
        "name": "ASP",
	      "number": 18,
	      "insertion_code": null,
        "sites":[17,18,19,20, ...]
      },
      {
        "name": "LEU",
	      "number": 18,
        "insertion_code": "A",
        "sites":[29,30,31, ...]
      },
    ]
  }

biomol_sequences
~~~~~~~~~~~~~~~~

> - **Description**: A list of residue sequences in current structure. It may be any type of sequence, as this type is further specified in :field:`biomol_sequence_types`.

Reviewer: Are you not duplicating data here? It seems that the sequences already occur in the biomol_chains field. It would perhaps be better to make a reference from the biomol_chains to these sequences.

Is the main point of this field not to enable querying on the sequences?
If that is the case, it may be better to set the query ability to SHOULD.

Author: You are right, data would be duplicated.

I was thinking that it could be better specifying which chains (and even which residues) are included on each sequence, by their indices, and then removing the _sequences_ and _sequence_types_ fields on _biomol_chains_.

Actually, instead of making new fields for this we could reshape sequences like a list of dictionaries, to make it coherent with the other biomol field formats. What do you think?

```rst

biomol_sequences
~~~~~~~~~~~~~~~~

- **Description**: A list of residue sequences in current structure.
 Every sequence is a dictionary which includes the sequence itself and the type of sequence it is.
 Every sequence may include a list of chain and residue indices.
 Sequences may be grouped and ordered in any form (e.g. by chains, by fragments of covalently bonded atoms, etc.) as long as they make sense when querying structures by sequence.
- **Type**: list of dictionaries with the properties:
   - :property:`sequence`: string (REQUIRED)
   - :property:`type`: string (REQUIRED)
   - :property:`chains`: list of integers
   - :property:`residues`: list of integers
- **Requirements/Conventions**:
   - **Query**:  Queries on this property SHOULD be supported.
   - **sequence**: A string with a letter for each residue in the sequence. Letters SHOULD be capital letters.
   - **type**: The type of a sequence is defined by its components (e.g. 'aminoacids').
   - **chains**: A list of integers referring to indices in :field:`biomol_chains` for chains which include this sequence totally or partially.
   - **residues**: A list of integers referring to indices in :field:`biomol_residues` for residues included in the sequence.
   Indices start the count at 0.
   There MUST NOT be repeated indices both in :property:`chains` and :property:`residues`.

- **Examples**:

.. code:: jsonc
  {
    "biomol_sequences":[
      {
        sequence: 'MSHHWGYG',
        type: 'aminoacids'
      },
      {
        sequence: 'GATTACA',
        type: 'nucleotides'
      }
    ]
  }

```

Reviewer: We could indeed turn the sequences into a dictionary.

I was still wondering whether there is a clear hierarchical structure? (That a chain can have one or more sequences, but a sequence can not contain more than one chain? In that case, we would not need to duplicate the residues. 

I was also wondering how the situation is handled, where there are multiple chains with the same sequence.
Will there be multiple sequence entries with the same sequence ? My idea was to make a sequence unique, so it only occurs once in the biomol_sequence field. But if you include the residues, you would need to have a separate sequence entry for each chain.

Author: > I was still wondering whether there is a clear hierarchical structure? (That a chain can have one or more sequences, but a sequence can not contain more than one chain? In that case, we would not need to duplicate the residues.

In my previous suggestion the chains property is a list of integers so a sequence may contain more than one chain. This is important since a polymer may be splitted in several chains. In the other hand, a chain may contain the whole sequence of a polymer and more things at the same time, so not all residues in the chain would be part of the sequence. Then listing residues in the chain makes sense.

> I was also wondering how the situation is handled, where there are multiple chains with the same sequence.
Will there be multiple sequence entries with the same sequence ? My idea was to make a sequence unique, so it only occurs once in the biomol_sequence field. But if you include the residues, you would need to have a separate sequence entry for each chain.

I did not think about this and you are right. Then we can forget about residues to make everything easier.

   - **sequences**: A list of residue sequences in current chain.
   - **sequence_types**: A list of tags specifying the type of each sequence in the :property:`sequences` field. The type of a sequence is defined by its components (e.g. 'aminoacids').
   - There SHOULD NOT be two or more chains with the same :property:`name`.
   - Values in :property:`name` SHOULD be in capital letters.
   - Values in :property:`name` SHOULD NOT be longer than 1 character when the number of chains is not greater than the number of letters in English alphabet (26).
   - Values in :property:`sequences` SHOULD be in capital letters.
   - Number of values in :property:`sequences` and :property:`sequence_types` MUST match.

- **Examples**:

.. code:: jsonc
  {
    "biomol_chains":[
      {
        "name": "A",
        "residues":[0,1,2,3, ...],
        "types": ['protein', 'ions'],
	      "sequences": ['MSHHWGYG'],
	      "sequence_types": ['aminoacids']
      },
      {
        "name": "B",
        "residues":[54,55,56,57, ...],
        "types": ['nucleic acid'],
	      "sequences": ['GATTACA'],
	      "sequence_types": ['nucleotides']
      },
    ]
  }

> biomol_residues

Reviewer: ```suggestion
_biomol_residues
```
I think the conclusion of the meeting last Friday was that it would be better to prepend the domain specific fields in the same way as the database providers prefix. So the names of the fields should prepended with an underscore.

Author: I tried to do the commit from your suggestion but I did not got it.
I did a new commit to change this.

Reviewer: That is ok, sometimes it is also easier to include it in your own branch first. I just wanted to make sure we would not forget to make the change.

   - **name**: The chain name/letter.
   - **residues**: A list of integers referring to the index of :field:`biomol_residues`, that belong to this chain.
     The list SHOULD NOT be empty. The index of the first residue is 0.
   - **types**: A list of custom tags/labels specifying the type of molecules this chain contains (e.g. 'protein').
     This field is useful as an overview of every chain and as a query target for the structure.

>      Labels in this field are non-standard. Every implementation may use different labels according to its needs.

Reviewer 2: I would advise standardizing labels at least for the most common molecule types to benefit the queryability of the field. Implementations could use their own labels, but prefixed with their own database-specific prefixes.

Author: Good idea. This way we could save a lot of work.

I have been searching for references for our labels in the current mmCIF format (the future PDB standard) and [this](https://mmcif.wwpdb.org/dictionaries/mmcif_em.dic/Items/_em_entity_assembly.type.html) is the best I have found. They are meant for assemblies and they do not totally suit me, but I will try to resemble them.

So I suggest the standard labels to be the following: 'PROTEIN', 'NUCLEIC ACID', 'CARBOHYDRATES', 'LIPID', 'MEMBRANE', 'LIGAND', 'ION', 'SOLVENT', 'OTHER'.

If you agree I will commit changes soon.

Reviewer: Yes, it seems like a good idea to define these labels.

Reviewer 2: I very much like basing on mmCIF. Maybe there is already a JSON representation for mmCIF data?

> So I suggest the standard labels to be the following: 'PROTEIN', 'NUCLEIC ACID', 'CARBOHYDRATES', 'LIPID', 'MEMBRANE', 'LIGAND', 'ION', 'SOLVENT', 'OTHER'.
> 
> If you agree I will commit changes soon.

These labels sound very good. I would just render them in lowercase and describe the use of prefixes for custom labels in the form of `<prefix>:<label>`.

Author: > Maybe there is already a JSON representation for mmCIF data?

In a fast search I found several mmCIF to JSON parsers and one of them seems to be the [official](https://pdbj.org/tools/cif2json) one.

> I would just render them in lowercase and describe the use of prefixes for custom labels in the form of <prefix>:<label>

Allright

- **Description**: For each residue in the system there is a dictionary that describes this residue. Residues are groups of related atoms (e.g. an aminoacid).
  Databases are allowed to add more properties as long as the properties are prefixed with the database specific prefix.
- **Type**: list of dictionaries with the properties:
   - :property:`name`: string (REQUIRED)
   - :property:`number`: integer (REQUIRED)
   - :property:`insertion_code`: string or null (REQUIRED)
   - :property:`chain`: string (OPTIONAL)
- **Requirements/Conventions**:
   - **Query**:  Support for queries on this property is OPTIONAL.
     If supported, only a subset of the filter features MAY be supported.
   - **name**: The residue name
   - **number**: The residue number according to source notation.
   - **insertion_code**: The residue insertion code. It MUST NOT be longer than 1 character. It MAY be null.
   - **chain**: The chain number this residue belongs to.
   - Values in :property:`chain` SHOULD be in capital letters.
   - Values in :property:`chain` SHOULD NOT be longer than 1 character when the number of chains is not greater than the number of letters in English alphabet (26).

>    - There MUST NOT be two or more residues with the same integer in :property:`sites`.

Reviewer: We no longer have a sites property so I think this line can go.

Author: ```suggestion
```

   - All :property:`name` and :property:`insertion_code` values SHOULD be in capital letters.

- **Examples**:

.. code:: jsonc
  {
    "_biomol_residues":[
      {
        "name": "PHE",
	"number": 17,
	"insertion_code": null,
        "chain": "A"
      },
      {
        "name": "ASP",
	"number": 18,
	"insertion_code": null,
        "chain": "A"
      },
      {
        "name": "LEU",
	"number": 18,
        "insertion_code": "A",
        "chain": "A"
      },
    ]
  }
  
_biomol_residues_at_sites
~~~~~~~~~~~~~~~~~~~~~~~~~

- **Description**: Index of the residues at each site (where values for sites are specified with the same order of the property `cartesian_site_positions`_).
  The properties of the residues are found in the property `_biomol_residues`_.
- **Type**: list of integers.
- **Requirements/Conventions**:

>   - **Support**: SHOULD be supported by all biomol implementations, i.e., SHOULD NOT be :val:`null`.

Author: ```suggestion
  - **Support**: MUST be supported when `_biomol_residues`_ is present as well, i.e., MUST NOT be :val:`null`.
```

  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.

>   - MUST have length equal to the number of sites in the structure (first dimension of the list property `cartesian_site_positions`_).

Reviewer: ```suggestion
  - The number of values MUST be equal to :property: `nsites`, i.e. the number of sites in the structure.
```

>   - Residue indices mentioned in the `_biomol_residues_at_sites`_ list MUST be lower than the length of the list property `_biomol_residues`_ (i.e. for each value in the `_biomol_residues_at_sites`_ list there MUST exist one dictionary in the `_biomol_residues`_ list with the index equal to the corresponding `_biomol_residues_at_sites`_ value).

Reviewer: ```suggestion
  - Each value in the `_biomol_residues_at_sites`_ list MUST correspond to  the index of one the dictionaries in the `_biomol_residues`_ list.
```
I think this point should be clear from the previous text. So I tried to shorten it. I am also ok with leaving this line out entirely.

Appendices
==========

Domain Specific Fields
----------------------

The fields below are all optional and are only used within specific research fields.

Every field has a standard domain-specific prefix.

_biomol_residues
~~~~~~~~~~~~~~~~

- **Description**: For each residue in the system there is a dictionary that describes this residue. Residues are groups of related atoms (e.g. an aminoacid).
  Databases are allowed to add more properties as long as the properties are prefixed with the database specific prefix.
- **Type**: list of dictionaries with the properties:
   - :property:`name`: string (REQUIRED)
   - :property:`number`: integer (REQUIRED)

>    - :property:`insertion_code`: string or null (REQUIRED)

Reviewer: Would icode be clear enough? It is a bit shorter than insertion_code.

  The properties of the residues are found in the property `_biomol_residues`_.
- **Type**: list of integers.
- **Requirements/Conventions**:
  - **Support**: MUST be supported when `_biomol_residues`_ is present as well, i.e., MUST NOT be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    If supported, filters MAY support only a subset of comparison operators.
  - The number of values MUST be equal to :property: `nsites`, i.e. the number of sites in the structure.
  - Each value in the `_biomol_residues_at_sites`_ list MUST correspond to  the index of one the dictionaries in the `_biomol_residues`_ list.

- **Examples**:

  - :val:`[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1, ... ]` indicates that the first 8 sites belong to the first residue in the the `residues`_ list, while the 9 following atoms belong to the second residue.

_biomol_site_sequences
~~~~~~~~~~~~~~~~~~~~~~

- **Description**: A list of dictionaries, each representing a linear segment of covalently-linked standard or modified amino acids or nucleotides having atoms with coordinates in sites. 

>   The order of the elements in the `_biomol_site_sequences`_ list is not relevant.

Reviewer: ```suggestion
  The elements in the `_biomol_site_sequences`_ list are unordered.
```
I think this has the same meaning and is a bit shorter.

  Each dictionary in the list holds two keys: sequence and type. The sequence is a string of one-letter codes identifying each amino acid or nucleotide as defined by the `mmCIF standard <https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_entity_poly.pdbx_seq_one_letter_code.html>`__.

>   The type is a string defining the monomers of the sequence. Accepted values are “polypeptide” for amino acids, “polydeoxyribonucleotide”  for deoxyribonucleotides and “polyribonucleotide” for ribonucleotides, according to the `mmCIF standard <https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_pdbx_reference_linked_entity.link_to_entity_type.html>`__.

Reviewer: ```suggestion
  The type is a string defining what kind of monomers are in the sequence.
    Accepted values are “polypeptide” for amino acids, “polydeoxyribonucleotide”  for deoxyribonucleotides and “polyribonucleotide” for ribonucleotides, according to the `mmCIF standard <https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v50.dic/Items/_pdbx_reference_linked_entity.link_to_entity_type.html>`__.
```

- **Type**: list of dictionaries with the properties:
   - :property:`sequence`: string (REQUIRED)
   - :property:`type`: string (REQUIRED)
- **Requirements/Conventions**:
   - **Query**:  Queries on this property SHOULD be supported.
   - **sequence**: A string with a letter for each residue in the sequence. Letters SHOULD be capital letters.
   - **type**: The type of a sequence is defined by the type of its residues (e.g. "polypeptide").

- **Examples**:

.. code:: jsonc
  {
    "_biomol_site_sequences":[
      {
        sequence: 'MSHHWGYG',
        type: 'polypeptide'
      },
      {

>         sequence: 'GATTACA',

Reviewer: ```suggestion
        sequence: '(DG)(DA)(DT)(DT)(DA)(DC)(DA)',
```
If this sequence represents DNA, should the DNA nucleotides not be encoded by two letter codes according to the mmCIF standard ?

        type: 'polydeoxyribonucleotide'
      }
    ]
  }
  
_biomol_full_sequences
~~~~~~~~~~~~~~~~~~~~~~

- **Description**: A list of dictionaries, each representing a linear segment of covalently-linked standard or modified amino acids or nucleotides including residues without coordinates in sites.

>   The order of the elements in the `_biomol_full_sequences`_ list is not relevant.

Reviewer: ```suggestion
  The elements in the `_biomol_full_sequences`_ list are unordered.
```

Appendices
==========

Domain Specific Fields
----------------------

The fields below are all optional and are only used within specific research fields.

Every field has a standard domain-specific prefix.

_biomol_residues
~~~~~~~~~~~~~~~~

- **Description**: For each residue in the system there is a dictionary that describes this residue. Residues are groups of related atoms (e.g. an aminoacid).
  Databases are allowed to add more properties as long as the properties are prefixed with the database specific prefix.
- **Type**: list of dictionaries with the properties:
   - :property:`name`: string (REQUIRED)
   - :property:`number`: integer (REQUIRED)
   - :property:`icode`: string or null (REQUIRED)

>    - :property:`chain`: string (OPTIONAL)

Reviewer 2: What does it mean when `chain` is missing?

Reviewer 2: Thanks for explanation. But maybe then it would make sense to make `chain` mandatory and faithfully retain the space character (` `) as its value?

Author: Sure, it also works to me.

We also talked about not getting constrained by the limits of PDB format regarding the 1 character string in the chain name, so the missing chain could also be 'Not defined', '', null or many others. As you prefer.
